lib/services/api/api_service.ts

```diff
@@ -234,4 +256,56 @@ export class SNApiService extends PureService {
 
     return response;
   }
+
+  private async checkForExpiredAccessToken(errorResponse: HttpResponse) {
+    if (errorResponse.status !== EXPIRED_ACCESS_TOKEN_RESPONSE_STATUS) {
+      return;
+    }
+    await this.refreshSession();
+  }
+
+  private async refreshSession() {
+    if (this.refreshingSession) {
+      return;
+    }
+    const user = await this.storageService!.getValue(StorageKey.User);
+    if (!user) {
+      return;
+    }
+    if (!user.uuid) {
+      return;
+    }
+    const userUuid = user.uuid;
+    this.refreshingSession = true;
+    const url = await this.path(REQUEST_PATH_SESSION_REFRESH);
+    const params = this.params({
+      user_uuid: userUuid,
+      access_token: this.session!.accessToken,
+      refresh_token: this.session!.refreshToken
+    });
+    const result = await this.httpService!.postAbsolute(
+      url,
+      params
+    ).then(async (response) => {
+      const session = this.newSessionFromResponse(response);
+      await this.storageService!.setValue(StorageKey.Session, session);
```

Comment:
Persisting a session should occur from either the api_service or the session_manager, not both. We can't be persisting the same object from two different locations. You can have session_manager rely on api_service instead to persist if you want.